File: flutter/lib/ui/painting/mask_filter.cc
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "flutter/lib/ui/painting/mask_filter.h"

#include "flutter/tonic/dart_args.h"
#include "flutter/tonic/dart_binding_macros.h"
#include "lib/tonic/converter/dart_converter.h"
#include "flutter/tonic/dart_library_natives.h"
#include "third_party/skia/include/effects/SkBlurMaskFilter.h"

namespace blink {

static void MaskFilter_constructor(Dart_NativeArguments args) {
  DartCallConstructor(&MaskFilter::Create, args);
}

IMPLEMENT_WRAPPERTYPEINFO(ui, MaskFilter);

void MaskFilter::RegisterNatives(DartLibraryNatives* natives) {
  natives->Register({
    { "MaskFilter_constructor", MaskFilter_constructor, 4, true },
  });
}

scoped_refptr<MaskFilter> MaskFilter::Create(unsigned style, double sigma,
                                             unsigned flags) {
  return new MaskFilter(SkBlurMaskFilter::Make(
      static_cast<SkBlurStyle>(style), sigma, flags));
}

MaskFilter::MaskFilter(sk_sp<SkMaskFilter> filter)
  : filter_(std::move(filter)) {
}

MaskFilter::~MaskFilter() {
}

} // namespace blink

File: lib/web_ui/lib/src/engine/canvaskit/mask_filter.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:ui/ui.dart' as ui;

import 'canvaskit_api.dart';
import 'skia_object_cache.dart';

/// The CanvasKit implementation of [ui.MaskFilter].
class CkMaskFilter extends ManagedSkiaObject<SkMaskFilter> {
  CkMaskFilter.blur(ui.BlurStyle blurStyle, double sigma)
      : _blurStyle = blurStyle,
        _sigma = sigma;

  final ui.BlurStyle _blurStyle;
  final double _sigma;

  @override
  SkMaskFilter createDefault() => _initSkiaObject();

  @override
  SkMaskFilter resurrect() => _initSkiaObject();

  SkMaskFilter _initSkiaObject() {
    return canvasKit.MaskFilter.MakeBlur(
      toSkBlurStyle(_blurStyle),
      _sigma,
      true,
    )!;
  }

  @override
  void delete() {
    rawSkiaObject?.delete();
  }
}
